### PR TITLE
New welcome flow

### PR DESCRIPTION
**Pull request by jedireza**
This PR implements the new welcome flow.

### How was this tested?
- [x] I tested this manually

### Manual test cases
Tested in the simulator. I plan on testing on my device also before landing.

### Screenshots and screen recordings
So many possibilities to screenshot. Test driving on different devices is recommended.

## Issues addressed
N/A

### CODE REVIEW

File: Client/Authentication/AuthStore.swift [hunk 1]
```diff
@@ -0,0 +1,454 @@
+// Copyright 2022 Neeva Inc. All rights reserved.
```
Comment (chung-lai) [inline Client/Authentication/AuthStore.swift:1]: Love this infra piece. Is the plan to switch `IntroViewModel` to use this AuthStore for authentication given it's mostly the same logic?

Comment (jedireza) [reply Client/Authentication/AuthStore.swift:1]: I was thinking that we could retire the whole `IntroViewModel` when we do a new "Sign Up/In" flow using the new stuff.

File: Client/Authentication/AuthStore.swift [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+import CryptoKit
+import Defaults
+import Shared
+
+private let log = Logger.auth
+
+class AuthStore: NSObject, ObservableObject {
+    public static let shared = AuthStore()
+
+    private var bvc = SceneDelegate.getBVC(for: nil)
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:15]: I don't think we should use `getBVC(for: nil)` here, especially since this view interacts with UI elements. Either should have it passed directly, or find a way to pass in/out only what's needed (my recommendation).

File: Client/Authentication/AuthStore.swift [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                .registerDeviceTokenWithServer(deviceToken: notificationToken)
+        }
+
+        let httpCookieStore = self.bvc.tabManager.configuration.websiteDataStore.httpCookieStore
+
+        httpCookieStore.setCookie(NeevaConstants.loginCookie(for: token))
+    }
+
+    private func openURLFromAuth(_ url: URL) {
+        if let selectedTab = self.bvc.tabManager.selectedTab,
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:34]: Please update this logic to use the change I made in #4144

File: Client/Authentication/AuthStore.swift [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        } else {
+            self.bvc.openURLInNewTab(url)
+            DispatchQueue.main.async {
+                self.bvc.hideCardGrid(withAnimation: false)
+            }
+        }
+    }
+
+    // MARK: - Auth Initializers
+
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:50]: nit: remove the extra line here

File: Client/Authentication/AuthStore.swift [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        else { return }
+
+        self.onError = onError
+        self.onSuccess = onSuccess
+
+        let session = ASWebAuthenticationSession(
+            url: authURL,
+            callbackURLScheme: NeevaConstants.neevaOAuthCallbackScheme()
+        ) { [self] callbackURL, error in
+
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:90]: nit: remove extra line

File: Client/Authentication/AuthStore.swift [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+            guard error == nil, let callbackURL = callbackURL else { return }
+            let queryItems = URLComponents(string: callbackURL.absoluteString)?.queryItems
+            let token = queryItems?.filter({ $0.name == "sessionKey" }).first?.value
+            let serverErrorCode = queryItems?.filter({ $0.name == "retry" }).first?.value
+
+            if let errorCode = serverErrorCode {
+                var errorMessage = "Some unknown error occurred"
+
+                switch errorCode {
+                case "NL003":
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:105]: Should these codes be hard coded in a shared location somewhere? Make them an `Enum` with `String` literals, so it's easier to read/reuse.

File: Client/Authentication/AuthStore.swift [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+                self.handleOauthWithProvider(token: cookie)
+            }
+        }
+
+        session.presentationContextProvider = self
+        session.start()
+    }
+
+    // MARK: - Auth Handlers
+
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:139]: Nit: remove extra line

File: Client/Authentication/AuthStore.swift [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+                    .registerDeviceTokenWithServer(deviceToken: notificationToken)
+            }
+        }
+
+        SpaceStore.shared.refresh(force: true)
+    }
+}
+
+// MARK: - Apple Auth Support
+
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:187]: ditto

File: Client/Authentication/AuthStore.swift [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+            if Defaults[.signedInOnce] {
+                ClientLogger.shared.logCounter(
+                    .SignInWithAppleSuccess,
+                    attributes: EnvironmentHelper.shared.getFirstRunAttributes()
+                )
+            }
+
+            self.handleSignUpWithApple(
+                identityToken: identityTokenStr, authorizationCode: authorizationCodeStr)
+            break
```
Comment (evancrow) [inline Client/Authentication/AuthStore.swift:235]: Can remove the `break` here. `break` is only needed if there is no code to execute for a `case`.

File: Client/Authentication/AuthStore.swift (hunk 1 above)
Comment (chung-lai) [reply Client/Authentication/AuthStore.swift:1]: Ah, makes sense. forgot we are redoing that flow as well. Exciting!

File: Client/Frontend/Overlay/OverlayManager.swift [hunk 10]
```diff
@@ -102,6 +103,22 @@ class OverlayManager: ObservableObject {
         show(overlay: .fullScreenModal(content), animate: animate, completion: completion)
     }
 
+    public func presentFullScreenCover(
```
Comment (evancrow) [inline Client/Frontend/Overlay/OverlayManager.swift:106]: The `FullScreenModal` is a full-screen cover, so `FullScreenCover` shouldn't be necessary. On iPad, it won't take up the full screen, but this is intentional to follow iPad Design Guidelines.

File: Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import Shared
+import SwiftUI
+
+struct WelcomeFlowPlansView: View {
+    @ObservedObject var model: WelcomeFlowModel
+
+    var premiumBullets = [
```
Comment (chung-lai) [inline Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift:11]: Should convert all these `String` type to `LocalizedStringKey` so they will be localized. Still in the middle of writing the doc about this :)

Comment (chung-lai) [reply Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift:11]: Ah, saw you are casting them to localized key but you will still need to add them to `Localizable.strings` file

File: Client/Authentication/AuthStore.swift (hunk 3 above)
Comment (jedireza) [reply Client/Authentication/AuthStore.swift:34]: This function might be going away in favor of letting the consumer choose what to do. In the welcome flow we don't want to open any tabs after authenticating. However, in a future sign up/in flow, we may indeed want to open a tab. Will revisit and ping you.

File: Client/Authentication/AuthStore.swift (hunk 6 above)
Comment (jedireza) [reply Client/Authentication/AuthStore.swift:105]: I expect this file to be the only location for these values, so I'm ok with them here like this.

File: Client/Frontend/Overlay/OverlayManager.swift (hunk 10 above)
Comment (jedireza) [reply Client/Frontend/Overlay/OverlayManager.swift:106]: > On iPad, it won't take up the full screen, but this is intentional to follow iPad Design Guidelines.

Yep and that's why I implemented this so we have full control when we need the overlay to be edge to edge.

File: Client/Authentication/AuthStore.swift (hunk 2 above)
Comment (jedireza) [reply Client/Authentication/AuthStore.swift:15]: 😅 I heard your voice when I put this line of code in. This isn't a view. If needed, we'll figure out a better way to get the bvc in here.

Comment (evancrow) [reply Client/Authentication/AuthStore.swift:15]: Haha, guess my voice wasn't very persuasive :). I think the BVC code is only used here for opening a new tab from sign-in, which shouldn't be a problem. I don't see the user having multiple windows when signing in. Also, hopefully, we can remove the need to open a new tab soon as well.

If we keep this, it should probably be a `let` instead of a `var`.

File: Client/Authentication/AuthStore.swift (hunk 3 above)
Comment (evancrow) [reply Client/Authentication/AuthStore.swift:34]: I figured we would remove this once we could inject the cookies natively without navigating to Neeva.com. Think the only hold-up was that OKTA was configured to work like this. (am I remembering correctly @chung-lai?)

File: Client/Frontend/Overlay/OverlayManager.swift (hunk 10 above)
Comment (evancrow) [reply Client/Frontend/Overlay/OverlayManager.swift:106]: Right, I'm just thinking we probably shouldn't use it because it goes against what Apple has designed for iPad. Also, if you want a full-width view, you can configure `OverlayPopoverView` via `OverlayStyle` to fill the view. This just adds another redundancy.

It's kind of confusing, would probably be good to have a quick doc explaining all the options (not asking you to write one, I should probably do that).

File: Client/Authentication/AuthStore.swift (hunk 2 above)
Comment (evancrow) [reply Client/Authentication/AuthStore.swift:15]: Also would change it to `getBVCOrNil()`, because it could be a bit more accurate in finding the active window if there are multiple.

File: Client/Authentication/AuthStore.swift (hunk 3 above)
Comment (chung-lai) [reply Client/Authentication/AuthStore.swift:34]: That's right..For OKTA flow, we need to open neeva.com in order to have the right user token set. https://github.com/neevaco/neeva-ios/issues/3614

I think it's because OKTA login was done in the SafariWebViewController and they have their own cookie jar and it doesn't communicate with our web view.

File: Client/Frontend/WelcomeFlow/WelcomeFlowModel.swift [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                         LogConfig.PromoCardAttribute.defaultBrowserInterstitialTrigger,
+                     value: trigger.rawValue
+                 )
+             ]
+         )
+         */
+    }
+
+    func complete() {
+        Defaults[.introSeen] = true
```
Comment (chung-lai) [inline Client/Frontend/WelcomeFlow/WelcomeFlowModel.swift:68]: Before, `introSeen` was set when user saw the welcome flow. In the new welcome flow, looks like we set this only when user completes the default browser flow or in SceneDelegate's `handleSignInToken` This makes sense to make sure user runs through the entire welcome flow. just curious if that's on purpose?

File: Client/Frontend/Premium/PremiumStore.swift [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+    init() {
+        Task {
+            self.loadingProducts = true
+            do {
+                self.products = try await Product.products(for: [
+                    PremiumPlan.monthly.rawValue, PremiumPlan.annual.rawValue,
+                ])
+            } catch {
+                // TODO: log
```
Comment (chung-lai) [inline Client/Frontend/Premium/PremiumStore.swift:40]: Do we plan to add the logging to the same PR? I would suggest we do another PR just for logging. Otherwise, it's a bit hard to see all the logging related changes.

**Review by chung-lai — COMMENTED** [review]
Left a few comments. Overall, the change looks good and love the refactoring! Would be great if we can break this huge PR into multiple smaller PRs to make it easier to digest but I guess it may be a little hard to pull them apart :)

File: Client/Frontend/WelcomeFlow/WelcomeFlowModel.swift (hunk 12 above)
Comment (jedireza) [reply Client/Frontend/WelcomeFlow/WelcomeFlowModel.swift:68]: Yes that was intentional.

File: Client/Frontend/Premium/PremiumStore.swift (hunk 13 above)
Comment (jedireza) [reply Client/Frontend/Premium/PremiumStore.swift:40]: We can do that if you prefer it.

File: Client/Frontend/WelcomeFlow/WelcomeFlowIntroView.swift [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                    }
+                )
+                .buttonStyle(.neeva(.secondary))
+            }
+
+            Spacer()
+
+            HStack {
+                Button(action: { collectUsageStats.toggle() }) {
+                    collectUsageStats
```
Comment (chung-lai) [inline Client/Frontend/WelcomeFlow/WelcomeFlowIntroView.swift:92]: We should call this when the user completes the flow

```
                Defaults[.shouldCollectUsageStats] = collectUsageStatsCheckbox
                if Defaults[.shouldCollectUsageStats] == true {
                    ClientLogger.shared.flushLoggingQueue()
                }
```

Comment (chung-lai) [reply Client/Frontend/WelcomeFlow/WelcomeFlowIntroView.swift:92]: This will determine we should enable logging for the entire app

**Comment by chung-lai** [comment]
Also, we may want to disable the first run test as we are changing the flow now.

File: Client/Authentication/AuthStore.swift (hunk 2 above)
Comment (jedireza) [reply Client/Authentication/AuthStore.swift:15]: Ok this has been reworked and the bvc is passed in explicitly. Thank you for calling this out.

File: Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+                            .foregroundColor(.brand.white)
+                            .frame(maxWidth: .infinity)
+                    }
+                )
+                .buttonStyle(.neeva(.primary))
+
+                if !NeevaUserInfo.shared.hasLoginCookie() {
+                    Button(
+                        action: {
+                            model.logCounter(.SignInClick)
```
Comment (chung-lai) [inline Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift:242]: We should log the screenName as an attribute for `SignInClick` logging so we can tell which screen's sign in button was tapped on.

**Review by chung-lai — APPROVED** [review]
One change on the SignInClick to add screen name. Other than that, LGTM with the logging and the usage stats change!

File: Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift (hunk 15 above)
Comment (chung-lai) [reply Client/Frontend/WelcomeFlow/WelcomeFlowPlansView.swift:242]: nvm, all the events will get the screenName as well
